source/extensions/tracers/opentelemetry/resource_detectors/environment/environment_resource_detector.cc

```diff
@@ -29,22 +29,21 @@ Resource EnvironmentResourceDetector::detect() {
   resource.schema_url_ = "";
   std::string attributes_str = "";
 
-  attributes_str = Config::DataSource::read(ds, true, context_.serverFactoryContext().api());
+  TRY_NEEDS_AUDIT {
+    attributes_str = Config::DataSource::read(ds, true, context_.serverFactoryContext().api());
+  }
+  END_TRY catch (const EnvoyException& e) {
+    ENVOY_LOG(warn, "Failed to detect resource attributes from the environment: {}.", e.what());
+  }
 
   if (attributes_str.empty()) {
-    throw EnvoyException(
-        fmt::format("The OpenTelemetry environment resource detector is configured but the '{}'"
-                    " environment variable is empty.",
-                    kOtelResourceAttributesEnv));
+    return resource;
   }
 
   for (const auto& pair : StringUtil::splitToken(attributes_str, ",")) {
     const auto keyValue = StringUtil::splitToken(pair, "=");
     if (keyValue.size() != 2) {
-      throw EnvoyException(
-          fmt::format("The OpenTelemetry environment resource detector is configured but the '{}'"
-                      " environment variable has an invalid format.",
-                      kOtelResourceAttributesEnv));
```

Comment:
I think this should still throw the exception because this should be treat as config error. We can accept the env variable absent but if it exist, it should have correct format?

cc @howardjohn WDYT?

/wait-any

Reply:
https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#parsing-empty-value seems to require a specific behavior?

Reply:
The spec above is indeed unclear (I will submit an issue for it), because I wasn't able to find a definition/behavior of "unset". But, I looked at all major SDKs resource implementation and **all** are treating unset/empty as no error (so, treat them as the same, which is what I expected), and returning an empty resource:

- [C++](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/sdk/src/resource/resource_detector.cc#L19)
- [Java](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ResourceConfiguration.java#L62C23-L62C42)
- [.NET](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Resources/OtelEnvResourceDetector.cs#L26)
- [PHP](https://github.com/open-telemetry/opentelemetry-php/blob/main/src/SDK/Resource/Detectors/Environment.php#L21-L23)
- [Go](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/resource/env.go#L54)
- [Collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/env/env.go#L45-L51)
- [JS](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-resources/src/detectors/EnvDetectorSync.ts#L57)

So that's why I wanted to keep the same behavior here.

Reply:
Another situation/use-case for empty could be: You set the `OTEL_RESOURCE_ATTRIBUTES` on the host with some attributes. But then on a specific app running on this host, you want to remove them. Then you would set for that specific app it empty. (or replace with something else..) So I think we shouldn't treat empty as an error here.

Reply:
Do I missed something? My initial comment is for the format error, seems you are discussing the empty value?

Reply:
Yeah, I was mostly replying to john's question. For invalid values I'd also not throw an exception. Maybe just log that it was invalid then. I think in general a resource detector should not fail due to this. Again, several other SDKs do the same:

- [.NET](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Resources/OtelEnvResourceDetector.cs#L43)
- [JS](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-resources/src/detectors/EnvDetectorSync.ts#L109)|
- [Python](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/resources/__init__.py#L316)

The OTel collector does return an [error from the detector ](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/env/env.go#L69-L71) but that error is [swallowed and simply logged ](- [Collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/resourcedetection.go#L130))

I just ran the collector with an invalid env value `OTEL_RESOURCE_ATTRIBUTES=extra=chars, a` and the collector starts but logs: 

```shell
otel-local-collector-1  | 2024-02-07T12:01:09.651Z      info    internal/resourcedetection.go:125       began detecting resource information    {"kind": "processor", "name": "resourcedetection/env", "pipeline": "traces"}
otel-local-collector-1  | 2024-02-07T12:01:09.654Z      warn    internal/resourcedetection.go:130       failed to detect resource       {"kind": "processor", "name": "resourcedetection/env", "pipeline": "traces", "error": "invalid resource format, invalid text: \" a\""}
otel-local-collector-1  | 2024-02-07T12:01:09.654Z      info    internal/resourcedetection.go:139       detected resource information   {"kind": "processor", "name": "resourcedetection/env", "pipeline": "traces", "resource": {}}
```

Reply:
Okay, if there are reasonable precedent.